book/src/internals/build-system.md

```diff
@@ -0,0 +1,100 @@
+
+### (Optional) Integration with CMake
+
+Somehow, all of this should be compatible with both CMake, and Cargo-only builds.
+
+## The plan (for now)
+
+After many rounds of refactoring this, we believe the best way to go is to not rely on cargos OUT_DIR exclusivly.
+Not being able to share artifacts between crates is very limiting.
+
+We want to use a similar approach to CXX, which stores its data within either:
+
+- Cargos `target/` directory under the `cxxbridge` subfolder.
+- A custom `scratch` directory
+
+See: [https://github.com/dtolnay/cxx/blob/afd4aa3f3d4e5d5e9a3a41d09df3408f5f86a469/gen/build/src/target.rs#L10](https://github.com/dtolnay/cxx/blob/afd4aa3f3d4e5d5e9a3a41d09df3408f5f86a469/gen/build/src/target.rs#L10) \
+and: [https://github.com/dtolnay/cxx/blob/afd4aa3f3d4e5d5e9a3a41d09df3408f5f86a469/gen/build/src/lib.rs#L178](https://github.com/dtolnay/cxx/blob/afd4aa3f3d4e5d5e9a3a41d09df3408f5f86a469/gen/build/src/lib.rs#L178).
+
+We will likely mirror this and use the `cxxqtbridge` naming.
+
+## Contents of `cxxqtbridge`
+
+### `crates` directory
+
+Inside cxxqtbridge, there should be a `crates` folder with one subfolder per crate.
+Each crates subfolder should contain the following:
+
+- `include/`
+  - `crate-name` - A folder for all headers that are exported by this crate
+  - `cxx-qt-lib -> ../../cxx-qt-lib/include` - Symbolic links for every dependency
+- `manifest.json` - This file describes which headers this library makes available, if it needs any Qt modules, etc.
+- `initializers.o` - The initializers of this crate + all it's dependencies to be linked in by CMake
+
+When building with cxx-qt-build, you may simply specify that your code depends on another crate.
+However, we also need to make sure that the order in which the build scripts are run works out, as e.g. the build script of cxx-qt-lib needs to run **before** any dependents build scripts run.
+
+For this use-case, Cargo has the `links` key (see the documentation [here](https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key)).
+It allows us to ensure the correct order, and also to pass metadata between build scripts.
+We specifically use this to let downstream dependents know where to find our manifest.json.
+Via the `manifest.json`, we are then able to figure out which header paths of this dependency to include, which Qt modules to link, etc.
+
+To make sure the correct data ends up in the manifest.json, we provide the `cxx_qt_build::Interface` struct which uses the builder pattern to specify all the necessary data.
+
+### `qml_modules` directory
+
+Next to the crates directory, there should be a `qml_modules` directory, which contains one directory per declared QML module.
+
+Each module should include a `plugin_init.o`, `.qmltypes`, `qmldir`, and any other necessary files.
+
+## Integration with CMake
+
+Via the `CXXQT_EXPORT_DIR` environment variable CMake should be able to change the location of the `cxxqtbridge` directory.
```

Comment:
We should probably have a separate issue to make sure our defines/env vars are consistent?

As i think we have `CXXQT_*` and `CXX_QT_*` ?

Reply:
```shell
cxx-qt on  985-build-system [$] is 📦 v0.6.1 via △ v3.27.7 via 🦀 v1.79.0
> rg CXXQT_ --vimgrep | wc -l
20

cxx-qt on  985-build-system [$] is 📦 v0.6.1 via △ v3.27.7 via 🦀 v1.79.0
> rg CXX_QT_ --vimgrep | wc -l
165
```
So it seems CXX_QT_ is the more dominant prefix...
We should probably track this in a separate issue though.

Reply:
Right could be a good first issue task ;-)

Reply:
Moved into #996